### PR TITLE
Test adc support temperature sensor stm32 mcu

### PR DESCRIPTION
**Pull request by TDhaouST**
The goal is to test the junction temperature sensor and the
vref int property in the ADC device-tree node of each stm32 where it is
available.

Check all the ADC of the stm32 mcus where the temp and VrefInt monitoring
is available (based on the Ref Man).
Check that has-temp-channel; and has-vref-channel; in the corresponding
ADC node of the DTS of each stm32 mcu is correctly set.
Verify the VTEMP/ VREFINT activation in the in adc_stm32.c for example).
It is necessary to add and activate the node die-temp in the boards dts files.
Execute the testcase running the samples/sensor/stm32_temp_sensor on
each available target board.

### CODE REVIEW

File: samples/sensor/stm32_temp_sensor/sample.yaml [hunk 1]
```diff
@@ -5,5 +5,10 @@ tests:
...
     tags: sensors
-    build_only: true
-    platform_allow: nucleo_f401re stm32f103_mini stm32l562e_dk
+    tags: tests
+    filter: dt_compat_enabled("st,stm32-temp-cal") or dt_compat_enabled("st,stm32-temp")
+    harness: console
+    harness_config:
+      type: one_line
+      regex:
+        - r"Current temperature ^([0-9]{1,2}){1}(\.[0-9]{1,2}) °C"
```
Comment (ABOSTM) [inline samples/sensor/stm32_temp_sensor/sample.yaml:14]: Modifications on sample part should be part of a separate commit.

I think that Regex could be simplified:
r"Current temperature [0-9]{1,2}\.[0-9]{1,2} °C"

**Review by ABOSTM — CHANGES_REQUESTED** [review]
Missing doc update for nucleo_g0b1 and stm32f030_demo

File: samples/sensor/stm32_temp_sensor/sample.yaml (hunk 1 above)
Comment (erwango) [reply samples/sensor/stm32_temp_sensor/sample.yaml:14]: +1 for the separate commit.
Regarding regex, let's limit expected t° range to 10>60.
Question on the "floating" part: Is it possible to get 2 digit ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/sensor/stm32_temp_sensor/sample.yaml (hunk 1 above)
Comment (TDhaouST) [reply samples/sensor/stm32_temp_sensor/sample.yaml:14]: According the example, the float part is composed of a single digit.

File: samples/sensor/stm32_temp_sensor/sample.yaml [hunk 2]
```diff
@@ -5,5 +5,10 @@ tests:
...
     tags: sensors
-    build_only: true
-    platform_allow: nucleo_f401re stm32f103_mini stm32l562e_dk
+    tags: tests
+    filter: dt_compat_enabled("st,stm32-temp-cal") or dt_compat_enabled("st,stm32-temp")
+    harness: console
+    harness_config:
+      type: one_line
+      regex:
+        - "Current temperature: [1-5][0-9].[0-9]{1} °C"
```
Comment (ABOSTM) [inline samples/sensor/stm32_temp_sensor/sample.yaml:14]:
```suggestion
        - "Current temperature: [1-5][0-9].[0-9] °C"
```

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: dts/arm/st/g4/stm32g4.dtsi [hunk 3]
```diff
@@ -86,6 +86,16 @@
 			has-temp-channel;
 		};
 
+		die_temp: dietemp {
+			label = "DIE_TEMP";
```
Comment (FRASTM) [inline dts/arm/st/g4/stm32g4.dtsi:90]: to be checked : stm32g4 has calibration

File: dts/arm/st/h7/stm32h7.dtsi [hunk 4]
```diff
@@ -736,6 +736,17 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
+		};
+
+		die_temp: dietemp {
+			label = "DIE_TEMP";
```
Comment (FRASTM) [inline dts/arm/st/h7/stm32h7.dtsi:744]: to be checked : stm32h7 has calibration (ex of stm32743)

File: drivers/sensor/stm32_temp/stm32_temp.c [hunk 5]
```diff
@@ -88,11 +88,37 @@ static int stm32_temp_channel_get(const struct device *dev, enum sensor_channel
...
+	temp *= cfg->cal2_temp - cfg->cal1_temp;
+	temp /= (*cfg->cal2_addr >> 2) - (*cfg->cal1_addr >> 2);
+	temp += cfg->cal_offset;
+#else
 	temp = ((float)data->raw * cfg->tsv_mv) / cfg->cal_vrefanalog;
 	temp -= *cfg->cal1_addr;
 	temp *= (cfg->cal2_temp - cfg->cal1_temp);
 	temp /= (*cfg->cal2_addr - *cfg->cal1_addr);
 	temp += cfg->cal_offset;
+#endif
```
Comment (erwango) [inline drivers/sensor/stm32_temp/stm32_temp.c:121]: Can we abstract this difference using device tree ?

Comment (FRASTM) [reply drivers/sensor/stm32_temp/stm32_temp.c:121]: Using the soc DTS  should introduce a new property <ADC resolution for TS_CAL> for each ADC with <st_stm32_temp_cal>.
The data TS_CAL1 (*cfg->cal1_addr) and TS_CAL2 (*cfg->cal2_addr) are stored in the system memory of the device. Depending on the SOC, the ADC is 12 or 14 or 16 bits wide and the TS_CAL1 and TS_CAL2 are measured at this max resolution. 
On the other hand the stm32_temp.c is always converting with 12bit resolution :
```
*asp = (struct adc_sequence){
		.resolution = 12U
}
```
That's why the shift is required.

File: drivers/sensor/stm32_temp/stm32_temp.c [hunk 6]
```diff
@@ -88,7 +88,20 @@ static int stm32_temp_channel_get(const struct device *dev, enum sensor_channel
 	}
 
 #if HAS_CALIBRATION
-#if defined(CONFIG_SOC_SERIES_STM32U5X)
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	/*
```
Comment (FRASTM) [inline drivers/sensor/stm32_temp/stm32_temp.c:93]: Please check that in details for stm32h7 devices: stm32h723,  stm32h735, stm32h7b3 etc. There are many ADC version in this soc family.

File: drivers/sensor/stm32_temp/stm32_temp.c (hunk 5 above)
Comment (FRASTM) [reply drivers/sensor/stm32_temp/stm32_temp.c:121]: Could also be done with a "custom flag" (like in  the  include/zephyr/dt-bindings/pwm/stm32_pwm.h with PWM_POLARITY_NORMAL  / PWM_STM32_COMPLEMENTARY)
--> TS_CAL_STM32_16BIT / TS_CAL_STM32_14BIT / TS_CAL_STM32_12BIT

Comment (erwango) [reply drivers/sensor/stm32_temp/stm32_temp.c:121]: > Could also be done with a "custom flag" (like in the include/zephyr/dt-bindings/pwm/stm32_pwm.h with PWM_POLARITY_NORMAL / PWM_STM32_COMPLEMENTARY) --> TS_CAL_STM32_16BIT / TS_CAL_STM32_14BIT as follows:
> 
> ```
> 		die_temp: dietemp {
> 			compatible = "st,stm32-temp-cal";
> 			ts-cal1-addr = <0x0BFA0710 | TS_CAL_STM32_16BIT>;
> 			ts-cal2-addr = <0x0BFA0742 | TS_CAL_STM32_16BIT>;
> ```

Using a flag on a 32 bits addr field looks a bit fragile. Can't we add a new `ts-cal-resolution` field ?

File: samples/sensor/stm32_temp_sensor/sample.yaml (hunk 2 above)
Comment (erwango) [reply samples/sensor/stm32_temp_sensor/sample.yaml:14]: @TDhaouST Can you take this request into account ?

File: dts/arm/st/f0/stm32f0.dtsi [hunk 7]
```diff
@@ -372,11 +372,23 @@
...
 			status = "disabled";
 			label = "ADC_1";
+			vref-mv = <3300>;
 			#io-channel-cells = <1>;
 			has-temp-channel;
 			has-vref-channel;
 		};
 
+		die_temp: dietemp {
+			label = "DIE_TEMP";
```
Comment (erwango) [inline dts/arm/st/f0/stm32f0.dtsi:382]: Nitpick: Please put "compatible" property first. "label" is a property that is close to be obsolete and is nearly not used. As such it feels strange seeing it first.

**Comment by pdietl** [comment]
Nice. I made a PR making _one_ of these changes.

File: dts/bindings/sensor/st,stm32-temp-cal.yaml [hunk 8]
```diff
@@ -43,3 +43,15 @@ properties:
       type: int
       required: true
       description: Temperature offset in Celcius
+
+    ts-cal-resolution:
+      type: int
+      required: false
+      description: |
+        temperature calibration resolution with which the ts-cal1-temp and
+        ts-cal2-temp are measured on 12, 14 and 16 bits.
```
Comment (erwango) [inline dts/bindings/sensor/st,stm32-temp-cal.yaml:52]:
```suggestion
        Temperature calibration resolution with which the ts-cal1-temp and
        ts-cal2-temp are measured.
```

File: drivers/sensor/stm32_temp/stm32_temp.c [hunk 9]
```diff
@@ -143,7 +144,8 @@ static const struct stm32_temp_config stm32_temp_dev_config = {
 	.cal1_temp = DT_INST_PROP(0, ts_cal1_temp),
 	.cal2_temp = DT_INST_PROP(0, ts_cal2_temp),
 	.cal_vrefanalog = DT_INST_PROP(0, ts_cal_vrefanalog),
-	.cal_offset = DT_INST_PROP(0, ts_cal_offset)
+	.cal_offset = DT_INST_PROP(0, ts_cal_offset),
+	.ts_cal_resolution = (DT_INST_PROP(0, ts_cal_resolution) - 12)
```
Comment (erwango) [inline drivers/sensor/stm32_temp/stm32_temp.c:148]: After doing the substraction `.ts_cal_resolution` should then be called `ts_cal_shift` as it's not a resolution anymore

```suggestion
	.ts_cal_shift = (DT_INST_PROP(0, ts_cal_resolution) - CAL_RES)
```
with
```
#define CAL_RES 12
```
(Prefer use of defines with explicit names rather than "magic"  numbers).
@ABOSTM, @FRASTM do you have an idea for a better name than `CAL_RES`

**Review by erwango — CHANGES_REQUESTED** [review]
Almost good! Some few more changes required.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/stm32_temp/stm32_temp.c (hunk 9 above)
Comment (FRASTM) [reply drivers/sensor/stm32_temp/stm32_temp.c]: RAW_DATA_RESOLUTION ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/sensor/st,stm32-temp-cal.yaml [hunk 10]
```diff
@@ -43,3 +43,15 @@ properties:
...
       required: true
       description: Temperature offset in Celcius
+
+    ts_cal_shift:
+      type: int
+      required: false
+      description: |
+        Temperature calibration resolution with which the ts-cal1-temp and
+        ts-cal2-temp are measured.
+      default: 12
```
[the remark this answers is not in the document]
Comment (erwango) [reply dts/bindings/sensor/st,stm32-temp-cal.yaml:50]: @MaureenHelm, are you good with this justification?

File: drivers/adc/adc_stm32.c [hunk 11]
```diff
@@ -853,11 +865,7 @@ static void adc_stm32_setup_speed(const struct device *dev, uint8_t id,
 #if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32L0X)
 	LL_ADC_SetSamplingTimeCommonChannels(adc,
 		table_samp_time[acq_time_index]);
-#elif defined(CONFIG_SOC_SERIES_STM32G0X)
-	/* Errata ES0418 and more: ADC sampling time might be one cycle longer */
-	if (acq_time_index  < 2) {
-		acq_time_index = 2;
-	}
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:843]: You should not remove this. It was just added recently. See https://github.com/zephyrproject-rtos/zephyr/pull/46125

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c [hunk 12]
```diff
@@ -93,14 +93,16 @@ static const uint32_t table_resolution[] = {
...
+	RES(14),
 #elif !defined(CONFIG_SOC_SERIES_STM32H7X)
 	RES(6),
 	RES(8),
 	RES(10),
 	RES(12),
-#if defined(CONFIG_SOC_SERIES_STM32U5X)
-	RES(14),
-#endif /* CONFIG_SOC_SERIES_STM32U5X */
 #else
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:106]: In a next PR we can use ts-cal-resolution to simplify this table.
Same with it's usage in `start_read()`

File: drivers/adc/adc_stm32.c [hunk 13]
```diff
@@ -607,9 +607,21 @@ static int start_read(const struct device *dev,
 	defined(CONFIG_SOC_SERIES_STM32WLX)
 	/* STM32G0 in "not fully configurable" sequencer mode */
 	LL_ADC_REG_SetSequencerChannels(adc, channel);
+	LL_ADC_REG_SetSequencerConfigurable(adc, LL_ADC_REG_SEQ_CONFIGURABLE);
```
Comment (FRASTM) [inline drivers/adc/adc_stm32.c:610]: this is going to change for stm32G0 too, but this device is preferably configured with 
LL_ADC_REG_SetSequencerConfigurable(adc, **LL_ADC_REG_SEQ_FIXED**);

File: dts/arm/st/g0/stm32g0.dtsi [hunk 14]
```diff
@@ -360,13 +360,26 @@
...
 			has-vbat-channel;
 		};
 
+		die_temp: dietemp {
+			compatible = "st,stm32-temp-cal";
+			ts-cal1-addr = <0x1FFF75A8>;
+			ts-cal2-addr = <0x1FFF75CA>;
+			ts-cal1-temp = <30>;
+			ts-cal2-temp = <130>;
+			ts-cal-vrefanalog = <3300>;
```
Comment (FRASTM) [inline dts/arm/st/g0/stm32g0.dtsi:377]: please check the Datasheet: 
"TS_CAL1 and TS_CAL2 are raw data acquired at a temperature of 30 °C  and 130 °C   (± 5 °C),  VDDA = VREF+ = **3.0 V** (± 10 mV)"

**Review by FRASTM — COMMENTED** [review]
running sample on the stm32g071rb nucleo gives : 
`Current temperature: 29.3 °C     `

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/sensor/st,stm32-temp-cal.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply dts/bindings/sensor/st,stm32-temp-cal.yaml:50]: Yes

**Review by MaureenHelm — DISMISSED** [review]

File: drivers/adc/adc_stm32.c [hunk 15]
```diff
@@ -583,7 +593,13 @@ static int start_read(const struct device *dev,
...
 	LL_ADC_SetChannelPreSelection(adc, channel);
 #endif
-
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+	/*
+	 * Each channel in the sequence must be previously enabled in PCSEL.
+	 * This register controls the analog switch integrated in the IO level.
+	 */
+	LL_ADC_SetChannelPreselection(adc, channel);
+#endif
```
[the remark this answers is not in the document]
Comment (TDhaouST) [reply drivers/adc/adc_stm32.c]: No it's not possible to factorize, i realized that it's not the same function LL_ADC_SetChannelPreselection for U5(Preselction) and LL_ADC_SetChannelPreSelection for H7 series (PreSelection).

[the remark this answers is not in the document]
Comment (erwango) [reply drivers/adc/adc_stm32.c:602]: Too bad, we need to raise an issue in Cube to get this fixed

**Review by erwango — DISMISSED** [review]

**Review by FRASTM — DISMISSED** [review]

**Review by MaureenHelm — DISMISSED** [review]

**Comment by MaureenHelm** [comment]
@ABOSTM please revisit

File: dts/arm/st/wb/stm32wb.dtsi [hunk 16]
```diff
@@ -357,12 +357,25 @@
...
+			vref-mv = <3000>;
 			label = "ADC_1";
 			#io-channel-cells = <1>;
 			has-temp-channel;
 			has-vref-channel;
 		};
 
+		die_temp: dietemp {
+			compatible = "st,stm32-temp-cal";
+			ts-cal1-addr = <0x1FFF75AA>;
```
Comment (ABOSTM) [inline dts/arm/st/wb/stm32wb.dtsi:369]: From STM32WB Datasheet, DS11929 Rev 12, address are wrong:
image

I didn't check all Datasheet, but please check all addresses.

File: dts/arm/st/l5/stm32l5.dtsi [hunk 17]
```diff
@@ -584,12 +584,25 @@
...
 			has-temp-channel;
 			has-vref-channel;
 		};
 
+		die_temp: dietemp {
+			compatible = "st,stm32-temp-cal";
+			ts-cal1-addr = <0x0BFA05A8>;
+			ts-cal2-addr = <0x0BFA05CA>;
+			ts-cal1-temp = <30>;
+			ts-cal2-temp = <110>;
```
Comment (ABOSTM) [inline dts/arm/st/l5/stm32l5.dtsi:599]: From datasheet, DS12737 Rev 6, Cal2 temp is 130°c not 110:
image

File: dts/arm/st/wb/stm32wb.dtsi [hunk 18]
```diff
@@ -357,12 +357,25 @@
...
+			ts-cal2-addr = <0x1FFF75AB>;
+			ts-cal1-temp = <30>;
+			ts-cal2-temp = <110>;
```
Comment (ABOSTM) [inline dts/arm/st/wb/stm32wb.dtsi:372]: `ts-cal2-temp` should be 130°c
Please check them all

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: dts/arm/st/f7/stm32f7.dtsi [hunk 19]
```diff
@@ -688,12 +688,23 @@
...
 			status = "disabled";
+			vref-mv = <3300>;
 			label = "ADC_1";
 			#io-channel-cells = <1>;
 			has-temp-channel;
 			has-vref-channel;
 		};
 
+		die_temp: dietemp {
+			compatible = "st,stm32-temp";
```
Comment (ABOSTM) [inline dts/arm/st/f7/stm32f7.dtsi:699]: dito

File: dts/arm/st/l4/stm32l4.dtsi [hunk 20]
```diff
@@ -371,10 +371,23 @@
...
 			label = "ADC_1";
 			#io-channel-cells = <1>;
 		};
 
+		die_temp: dietemp {
+			compatible = "st,stm32-temp-cal";
+			ts-cal1-addr = <0x1FFF75A8>;
+			ts-cal2-addr = <0x1FFF75CA>;
+			ts-cal1-temp = <30>;
+			ts-cal2-temp = <130>;
```
Comment (ABOSTM) [inline dts/arm/st/l4/stm32l4.dtsi:384]:
```suggestion
			ts-cal2-temp = <110>;
```
`Please check them all`

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: dts/arm/st/h7/stm32h7.dtsi [hunk 21]
```diff
@@ -676,6 +676,20 @@
...
+
+		die_temp: dietemp {
+			compatible = "st,stm32-temp-cal";
+			ts-cal1-addr = <0x1FF1E820>;
+			ts-cal2-addr = <0x1FF1E840>;
+			ts-cal1-temp = <30>;
+			ts-cal2-temp = <110>;
+			ts-cal-vrefanalog = <3300>;
+			ts-cal-resolution = <16>;
+			io-channels = <&adc3 18>;
```
Comment (ABOSTM) [inline dts/arm/st/h7/stm32h7.dtsi:691]: Unfortunately, channel is not the same for H7 series:
for example H723/H725/H730 is channel 17
Please check channel for all H7 MCU.

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: dts/arm/st/h7/stm32h723.dtsi [hunk 22]
```diff
@@ -30,6 +30,10 @@
 			status = "disabled";
 		};
 
+		die_temp: dietemp {
+			io-channels = <&adc3 17>;
```
Comment (ABOSTM) [inline dts/arm/st/h7/stm32h723.dtsi:34]: H723 has TS_CAL2 130°c

File: dts/arm/st/h7/stm32h7a3.dtsi [hunk 23]
```diff
@@ -58,6 +58,10 @@
 			label= "OCTOSPI1";
 		};
 
+		die_temp: dietemp {
+			io-channels = <&adc2 18>;
```
Comment (ABOSTM) [inline dts/arm/st/h7/stm32h7a3.dtsi:62]: H7A3 has TS_cal2 130°c
also address are different

**Review by ABOSTM — CHANGES_REQUESTED** [review]

**Review by ABOSTM — DISMISSED** [review]
LGTM

**Review by FRASTM — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Review by FRASTM — DISMISSED** [review]
LGTM 
but should this PR about ADC temp sensor mix commits 1e614d2c1701a41e462190a177b305d206e696b0 and eaf09a885bb07f30501d48f5c241dad84e4e723d about _drivers: adc: stm32 init the Regular group_

**Comment by carlescufi** [comment]
@TDhaouST please rebase

**Review by ABOSTM — APPROVED** [review]

**Review by erwango — APPROVED** [review]
